class-plugin-autoupdate-filter.php

```diff
@@ -14,29 +14,104 @@ class Plugin_Autoupdate_Filter {
+		add_filter( 'plugin_auto_update_setting_html', array( $this, 'custom_setting_html' ), 11, 3 );
 
 		//Append text to upgrade text on plugins page for plugins explicitly set to not autoupdate
-		add_action( 'admin_init', array( $this, 'plugin_autoupdate_filter_change_upgrade_message_for_specific_plugins' ) );
+		add_action( 'admin_init', array( $this, 'upgrade_message_for_specific_plugins' ) );
 
 		// Always send auto-update emails to T51 concierge email address
-		add_filter( 'auto_plugin_theme_update_email', array( $this, 'plugin_autoupdate_filter_custom_update_emails' ), 10, 4 );
-		add_filter( 'auto_core_update_email', array( $this, 'plugin_autoupdate_filter_custom_update_emails' ), 10, 4 );
-		add_filter( 'automatic_updates_debug_email', array( $this, 'plugin_autoupdate_filter_custom_debug_email' ), 10, 3 );
+		add_filter( 'auto_plugin_theme_update_email', array( $this, 'custom_update_emails' ), 10, 4 );
+		add_filter( 'auto_core_update_email', array( $this, 'custom_update_emails' ), 10, 4 );
+		add_filter( 'automatic_updates_debug_email', array( $this, 'custom_debug_email' ), 10, 3 );
 
 		// re-enable core update emails which are disabled in an mu-plugin at the Atomic platform level
 		add_filter( 'automatic_updates_send_debug_email', '__return_true', 11 );
 		add_filter( 'auto_core_update_send_email', '__return_true', 11 );
 		add_filter( 'auto_plugin_update_send_email', '__return_true', 11 );
 		add_filter( 'auto_theme_update_send_email', '__return_true', 11 );
 
+		// "Disable all autoupdates" toggle
+		add_filter( 'auto_update_plugin', array( $this, 'maybe_disable_all_autoupdates' ), PHP_INT_MAX, 2 );
+		add_filter( 'auto_update_core', array( $this, 'maybe_disable_all_autoupdates' ), PHP_INT_MAX, 2 );
+		add_filter( 'auto_update_theme', array( $this, 'maybe_disable_all_autoupdates' ), PHP_INT_MAX, 2 );
+		add_action( 'admin_init', array( $this, 'autoupdate_settings_admin_notice' ) );
+
+	}
+
+	/**
+	 * Load settings from the centralized settings page
+	 */
+	private function get_auto_update_settings(): stdClass|WP_Error {
+
+		$response = wp_remote_get(
+			'https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/',
+			array( 'headers' => array( 'Accept' => 'application/json' ) )
+		);
+
+		if ( is_wp_error( $response ) ) {
+			throw new RuntimeException( $response->get_error_message(), $response->get_error_code() );
+		}
+
+		$response_code = wp_remote_retrieve_response_code( $response );
+		$response_body = wp_remote_retrieve_body( $response );
+
+		// Check that the response code is a 2xx code.
+		if ( ! \str_starts_with( (string) $response_code, '2' ) ) {
+			$response_message = wp_remote_retrieve_response_message( $response );
+			throw new Exception( $response_message, $response_code );
+		}
+
+		$decoded_body = json_decode( $response_body, false, 512, JSON_THROW_ON_ERROR );
+
+		// if the settings are empty, we still need to return an object
```

Comment:
If `$response_body` is empty, then `json_decode` will throw an error and it will be caught. You can try it out yourself here: https://onlinephp.io/c/df536

Then, you set the function return type to be `stdClass|WP_Error`. If the JSON deserializes to something else, like an array, PHP itself will throw an exception! You can try that here: https://onlinephp.io/c/88fcd

So basically, you can immediately return line 88. If `$response_body` contains anything other than a JSON object, line 22 will catch that. Easy error handling! 😄

Reply:
I've tried this several times, and if I don't cast it specifically as an object, this is the error I get:

`PHP Fatal error:  Uncaught TypeError: Plugin_Autoupdate_Filter::get_auto_update_settings(): Return value must be of type stdClass, string returned in /Users/nickgreen/Local Sites/testi/app/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:101`

To reproduce:
1. Toggle off the "disable all autoupdates" toggle
2. See that the JSON being returned from the site is `""`
3. Get that return value error, because that `json_decode` line is returning a string 🤔

Reply:
The fix is in line 22. Change `catch ( Exception $exception )` to `catch ( Exception|Error $exception )`.

**Or** simply change it to `catch ( Throwable $e )` since both the [`Exception`](https://www.php.net/manual/en/class.exception.php) and [`Error`](https://www.php.net/manual/en/class.error.php) classes in PHP implement [the `Throwable` interface](https://www.php.net/manual/en/class.throwable.php).

Reply:
Based on the docs, an `Error` is the base class for all internal PHP errors(like the returned type being wrong), whereas `Exception` is the base class for all user errors.

Reply:
The problem here is that it shouldn't be an error at all. 

When I toggle the switch off in the settings (meaning no settings are actually set), then it returns an empty string there. But this isn't an indication of an error, and I don't want the function to treat it as an error. I want it to interpret it as "No error, but no settings are set". 

Anyway, we can fix this in a future iteration, and I'm interested in getting something out the door soon, so I'm going to stick a pin in it for later. :)

Reply:
> and I don't want the function to treat it as an error

I agree completely! 

> then it returns an empty string there

I would consider this a bug on OpsOasis' end though. If the switch is turned off, you'd still expect a valid object as a return value. But the setting is set to "false" instead of "true".

Reply:
Good point, maybe I'm looking at this from the wrong end entirely, and should be making sure that the data returned from opsoasis is correct.